lib/Sema/TypeCheckType.cpp

```diff
@@ -1979,6 +1974,16 @@ Type ResolveTypeRequest::evaluate(Evaluator &evaluator,
   if (validateAutoClosureAttributeUse(ctx.Diags, TyR, result, options))
     return ErrorType::get(ctx);
 
+  // Diagnose an attempt to use a placeholder at the top level.
+  if (isa<PlaceholderTypeRepr>(TyR) &&
+      resolution->getOptions().contains(TypeResolutionFlags::Direct)) {
+    if (!resolution->getOptions().contains(TypeResolutionFlags::SilenceErrors))
+      ctx.Diags.diagnose(loc, diag::top_level_placeholder_type);
```

Comment:
Should this come with a fix-it to remove the `_`? Or perhaps a note+fix-it saying something along the lines of `remove '_' to infer the top-level type?`?

Reply:
Hmm, yeah, I think that makes sense in most cases. We'd need to remove more than just the underscore (e.g. "`: _`" in `let x: _ = 0` or "` as _`" in `0 as _`), but some situations are a bit more complex—e.g. we probably want to turn `_()` into `.init()` since removing `_` produces nonsense. In any case we'd have to move this check outside of the type resolver to a place where we have a bit more context... 🤔

Reply:
Ahh good point. I think the fix-it is less important than making the message more actionable, but I'm not quite sure how to word it